File: 310971/Stankevich/rest310/src/main/java/com/task/rest/repository/InMemoryRepository.java
package com.task.rest.repository;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

public abstract class InMemoryRepository<T> implements CrudRepository<T, Long> {

    protected final Map<Long, T> storage = new ConcurrentHashMap<>();
    protected final AtomicLong idGenerator = new AtomicLong(1);

    protected abstract Long getId(T entity);
    protected abstract void setId(T entity, Long id);

    @Override
    public T save(T entity) {
        if (getId(entity) == null) {
            setId(entity, idGenerator.getAndIncrement());
        }
        storage.put(getId(entity), entity);
        return entity;
    }

    @Override
    public Optional<T> findById(Long id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public List<T> findAll() {
        return new ArrayList<>(storage.values());
    }

    @Override
    public T update(T entity) {
        Long id = getId(entity);
        if (id == null || !storage.containsKey(id)) {
            throw new IllegalArgumentException("Entity not found with id: " + id);
        }
        storage.put(id, entity);
        return entity;
    }

    @Override
    public void deleteById(Long id) {
        storage.remove(id);
    }

    @Override
    public boolean existsById(Long id) {
        return storage.containsKey(id);
    }
}

File: 351002/Bubich/t310/app/repository/in_memory_repository.py
from typing import List, Optional, Dict, TypeVar, Generic
from app.repository.base_repository import BaseRepository

T = TypeVar('T')


class InMemoryRepository(BaseRepository, Generic[T]):
    def __init__(self):
        self._storage: Dict[int, T] = {}
        self._next_id = 1

    def save(self, entity: T) -> T:
        entity.id = self._next_id
        self._storage[self._next_id] = entity
        self._next_id += 1
        return entity

    def find_by_id(self, id: int) -> Optional[T]:
        return self._storage.get(id)

    def find_all(self) -> List[T]:
        return list(self._storage.values())

    def update(self, entity: T) -> T:
        if entity.id not in self._storage:
            raise ValueError(f"Entity with id {entity.id} not found")
        self._storage[entity.id] = entity
        return entity

    def delete_by_id(self, id: int) -> bool:
        if id in self._storage:
            del self._storage[id]
            return True
        return False